File: scripts/notion_to_content/yamlfmt.py
from __future__ import annotations

import json
from typing import Any, Mapping


def _quote(s: str) -> str:
    return json.dumps(s, ensure_ascii=False)


def _scalar(v: Any) -> str:
    if v is None:
        return "null"
    if isinstance(v, bool):
        return "true" if v else "false"
    if isinstance(v, int) and not isinstance(v, bool):
        return str(v)
    if isinstance(v, float):
        if v != v:  # nan
            return "null"
        return repr(v) if (v + 0.0) != int(v) else str(int(v))
    if isinstance(v, str):
        return _quote(v)
    return _quote(str(v))


def format_front_matter(data: Mapping[str, Any]) -> str:
    """Emit YAML front matter without PyYAML: shallow dict, flat lists, scalars."""
    lines: list[str] = ["---"]
    for k, v in data.items():
        if v is None:
            continue
        if isinstance(v, (list, tuple)):
            lines.append(f"{k}:")
            for item in v:
                lines.append(f"  - {_quote(str(item))}")
        else:
            lines.append(f"{k}: {_scalar(v)}")
    lines.append("---")
    return "\n".join(lines) + "\n"

File: packages/notion-importer/src/yamlfmt.ts
function quote(s: string): string {
  return JSON.stringify(s);
}

function scalar(v: unknown): string {
  if (v === null || v === undefined) return "null";
  if (typeof v === "boolean") return v ? "true" : "false";
  if (typeof v === "number") {
    if (Number.isNaN(v)) return "null";
    return Number.isInteger(v) ? String(v) : String(v);
  }
  if (typeof v === "string") return quote(v);
  return quote(String(v));
}

/** Emit YAML front matter without a YAML library: shallow dict, flat lists, scalars. */
export function formatFrontMatter(data: Record<string, unknown>): string {
  const lines: string[] = ["---"];
  for (const [k, v] of Object.entries(data)) {
    if (v === null || v === undefined) continue;
    if (Array.isArray(v)) {
      lines.push(`${k}:`);
      for (const item of v) {
        lines.push(`  - ${quote(String(item))}`);
      }
    } else {
      lines.push(`${k}: ${scalar(v)}`);
    }
  }
  lines.push("---");
  return `${lines.join("\n")}\n`;
}
